### ISSUE TITLE
Magicsock: optimize reception of UDP packets from derper on linux

### ISSUE THREAD

**Issue by rklaehn** [n0-computer/iroh#1358]
When udp packets are received from the derper, on linux there is the possibility of 

a) using multiple slots in poll_recv
b) using multiple udp packets per slot

While we use a), we don't use b). We just put 1 packet into each slot and set the segment_size to the size of the entire packet. I have some doubts that this works at all, but even if it works it is less than optimal.

Each slot should be filled as much as possible before moving on to the next slot - that way the total number of udp packets you can handle in a single poll_recv call is maximized.

**Comment by rklaehn** [comment]
Related to https://github.com/n0-computer/iroh/issues/1357

**Comment by flub** [comment]
I think in general we might be losing the GSO optimsation on the send side.  Which may be the same as you're describing here.  I wanted to create an issue but need to check if this is really true first.

**Comment by rklaehn** [comment]
I doubt this is very critical perf wise.

But in principle on the linux side you could suck tons of UDP packets from the TCP stream and then fill a **lot** of slots in a single poll_recv call. There are 32 slots, and each slot can hold up to ~80 UDP packets.

**Comment by flub** [comment]
> I doubt this is very critical perf wise.

Maybe because you're on mac and there you never get segmentation offloading... But on linux this is know to have a huge perf impact and segmentation offloading is required to get speeds that are in the ballpark of TCP.

**Comment by rklaehn** [comment]
> > I doubt this is very critical perf wise.
> 
> Maybe because you're on mac and there you never get segmentation offloading... But on linux this is know to have a huge perf impact and segmentation offloading is required to get speeds that are in the ballpark of TCP.

Yeah, I mean it is not perf critical because once you use the derper perf is not going to be that great anyway... :-)

**Comment by rklaehn** [comment]
There is something that causes the derper to constantly disconnect. We need to fix this before even thinking about doing this issue... https://github.com/n0-computer/iroh/issues/1381

**Comment by dignifiedquire** [comment]
what is left here to do?
